tests/pipeline_test.py

```diff
@@ -577,11 +577,114 @@ class B(Generic[T]):
         pl[B[int]] = 1.0
 
 
-def test_setitem_raises_if_key_exists() -> None:
+def test_setitem_can_replace_param_with_param() -> None:
     pl = sl.Pipeline()
     pl[int] = 1
-    with pytest.raises(ValueError):
-        pl[int] = 2
+    pl[int] = 2
+    assert pl.compute(int) == 2
+
+
+def test_insert_can_replace_param_with_provider() -> None:
+    def func() -> int:
+        return 2
+
+    pl = sl.Pipeline()
+    pl[int] = 1
+    pl.insert(func)
+    assert pl.compute(int) == 2
+
+
+def test_setitem_can_replace_provider_with_param() -> None:
+    def func() -> int:
+        return 2
+
+    pl = sl.Pipeline()
+    pl.insert(func)
+    pl[int] = 1
+    assert pl.compute(int) == 1
+
```

Comment:
```suggestion

def test_setitem_can_replace_instance() -> None:
    pl = sl.Pipeline()
    original_number = 1
    pl[int] = original_number
    assert pl.compute(int) is original_number
    new_number = 1
    pl[int] = new_number
    assert pl.compute(int) == original_number
    assert pl.compute(int) is not original_number
    assert pl.compute(int) is new_number

```

Can we also have a test case that shows instance check like this...?
Just to show that it's passed by lambda function, so the instance is kept not just the value.

Reply:
`int` is immutable, not sure I see the value of such a test?

Reply:
Yes, but it shows that it doesn't copy the value and return it but keep the instance itself.
In case we want to pass around ``dict``, it might be useful information I thought...?

I just suggested what I wanted to try with the changes.
Feel free to drop them...!